File: core/src/main/java/com/natchuz/hub/core/proxy/NetworkProxyBackend.java
package com.natchuz.hub.core.proxy;

import org.bukkit.entity.Player;

import com.natchuz.hub.protocol.arch.Services;
import com.natchuz.hub.protocol.messaging.Protocol;
import com.natchuz.hub.protocol.state.JoinFlags;

public class NetworkProxyBackend implements ProxyBackend {

    private final Protocol protocol;

    public NetworkProxyBackend(Protocol protocol) {
        this.protocol = protocol;
    }

    @Override
    public void send(String name, String target, JoinFlags... flags) {
        String[] params = new String[2 + flags.length];
        params[0] = name;
        params[1] = target;

        for (int i = 0; i < flags.length; i++)
            params[i + 2] = flags[i].toString();

        protocol.send(Services.BUNGEECORD.messageEndpoint(), "send", params);
    }

    @Override
    public void send(Player player, String target, JoinFlags... flags) {
        send(player.getName(), target, flags);
    }
}

File: core/network/src/main/kotlin/com/natchuz/hub/core/network/NetworkProxyBackend.kt
package com.natchuz.hub.core.network

import com.natchuz.hub.backend.state.PlayerFlags
import com.natchuz.hub.backend.state.SendRequest
import com.natchuz.hub.core.api.proxy.ProxyBackend
import io.ktor.client.*
import io.ktor.client.request.*
import kotlinx.coroutines.runBlocking
import org.spongepowered.api.Sponge
import org.spongepowered.api.entity.living.player.Player

class NetworkProxyBackend(
        private val client: HttpClient
) : ProxyBackend {

    override fun send(name: String, target: String, vararg flags: PlayerFlags) {
        send(Sponge.getServer().getPlayer(name).get(), target, *flags)
    }

    override fun send(player: Player, target: String, vararg flags: PlayerFlags) = runBlocking {
        client.post<Unit>("http://state/player/${player.uniqueId}/send",
                body = SendRequest(target, flags.asList()))
    }
}